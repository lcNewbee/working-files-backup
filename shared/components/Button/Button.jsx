import React from 'react';
import PropTypes from 'prop-types';
import PureComponent from '../Base/PureComponent';
import Icon from '../Icon';
import utils from '../../utils';

// Styles
import '../../scss/01_atom/_a-button.scss';

const propTypes = {
  icon: PropTypes.string,
  className: PropTypes.string,
  theme: PropTypes.oneOf(['primary', 'secondary', 'success', 'info', 'warning', 'danger']),
  size: PropTypes.oneOf(['sm', 'lg', 'min']),
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
  inverse: PropTypes.bool,
  Component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  loading: PropTypes.bool,
  text: PropTypes.string,
  onClick: PropTypes.func,
};

const defaultProps = {
  Component: 'button',
  theme: 'secondary',
  role: 'button',
};

class Button extends PureComponent {
  constructor(props) {
    super(props);

    utils.binds(this, ['onClick']);
    this.state = {
      clicked: false,
    };
  }
  componentWillUnmount() {
    clearTimeout(this.clickTimeout);
  }

  onClick(e) {
    // loading 状态不响应点击事件
    if (!this.props.loading && this.props.onClick) {
      this.props.onClick(e);
      this.setState({
        clicked: true,
      });
      this.clickTimeout = setTimeout(() => {
        this.setState({
          clicked: false,
        });
      }, 500);
    }
  }

  render() {
    const { Component, icon, size, theme, className, loading, text, type, inverse } = this.props;
    const componentProps = utils.extend({}, this.props);
    const myIcon = icon ? <Icon name={icon} /> : null;

    let classNames = 'a-btn';

    if (size) {
      classNames = `${classNames} a-btn--${size}`;
    }

    if (theme) {
      classNames = `${classNames} a-btn--${theme}`;
    }

    if (!text) {
      classNames = `${classNames} a-btn--no-text`;
    }

    if (className) {
      classNames = `${classNames} ${className}`;
    }
    if (inverse) {
      classNames = `${classNames} a-btn--inverse`;
    }
    if (loading) {
      classNames = `${classNames} a-btn--loading`;
    }

    if (this.state.clicked) {
      classNames = `${classNames} a-btn--clicked`;
    }

    if (Component === 'button' || Component === 'input') {
      delete componentProps.text;
      delete componentProps.savedText;
      delete componentProps.savingText;
      delete componentProps.Component;
      delete componentProps.loading;
      delete componentProps.theme;
      delete componentProps.actionName;
      delete componentProps.needConfirm;

      componentProps.type = type;
    }

    return (
      <Component
        {...componentProps}
        className={classNames}
        onClick={this.onClick}
      >
        {
          loading ? (
            <Icon name="spinner" spin />
          ) : myIcon
        }
        {text}
      </Component>
    );
  }
}

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;
