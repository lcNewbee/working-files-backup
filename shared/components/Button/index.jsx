import React, { PropTypes } from 'react';
import PureRenderMixin from 'react-addons-pure-render-mixin';
import Icon from '../Icon';
import utils from '../../utils';
import './_button.scss';

const propTypes = {
  iconName: PropTypes.string,
  icon: PropTypes.string,
  className: PropTypes.string,
  role: PropTypes.oneOf(['default', 'primary', 'success', 'info', 'warning', 'danger']),
  size: PropTypes.oneOf(['sm', 'lg']),
  inverse: PropTypes.bool,
  Component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  loading: PropTypes.bool,
  text: PropTypes.string,
};

const defaultProps = {
  Component: 'button',
  role: 'default',
};

class Button extends React.Component {
  constructor(props) {
    super(props);

    this.shouldComponentUpdate = PureRenderMixin.shouldComponentUpdate.bind(this);
  }

  render() {
    const { Component, icon, size, role, className, loading, text } = this.props;
    const componentProps = utils.extend({}, this.props);

    let classNames = 'btn';

    if (size) {
      classNames = `${classNames} btn-${size}`;
    }

    if (role) {
      classNames = `${classNames} btn-${role}`;
    }

    if (className) {
      classNames = `${className} ${classNames}`;
    }

    if (Component === 'button' || Component === 'input') {
      delete componentProps.text;
      delete componentProps.Component;
      delete componentProps.loading;
    }

    return (
      <Component
        {...this.props}
        className={classNames}
        type="button"
      >
        {
          loading ? (
            <Icon name="spinner" spin />
          ) : <Icon name={icon} />
        }
        {text}
      </Component>
    );
  }
}

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;
