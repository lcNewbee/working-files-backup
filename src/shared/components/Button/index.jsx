import './index.scss';
import React, {PropTypes} from 'react';
import Icon from '../Icon';

const propTypes = {
  iconName: PropTypes.string,
  icon: PropTypes.string,
  className: PropTypes.string,
  role: PropTypes.oneOf(['default', 'primary', 'success', 'info', 'warning']),
  size: PropTypes.oneOf(['sm', 'lg']),
  inverse: PropTypes.bool,
  Component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
};

const defaultProps = {
  Component: 'button',
  role: 'default'
};

class Button extends React.Component {

  render() {
    let { Component, icon, size, role, className, loading, text} = this.props;
    
    let classNames = `btn`;
    
    if (size) {
      classNames = `${classNames} btn-${size}`;
    }
    
    if (role) {
      classNames = `${classNames} btn-${role}`;
    }
    
    if (className) {
      classNames = className + ' ' + classNames;
    }
    
    return (
      <Component 
        {...this.props}
        className={classNames}
        type="button"
      >
        {
          loading ? (
            <Icon name="spinner" spin={true} style={{marginLeft: '3px'}}/>
          ) : <Icon name={icon} />
        }
        {text}
      </Component>
    );
  }
}

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;