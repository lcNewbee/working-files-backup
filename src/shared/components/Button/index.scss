@import "../../scss/variables";
@import "../../scss/mixins";

//
// Button
// --------------------------------------------------

.btn {
  display: inline-block;
  padding: 6px 12px;
  font-size: $g-font-size;
  color: #666;
  text-decoration: none;
  background-color: #e8e8e8;
  background-image: linear-gradient(#e8e8e8,#e2e2e2);
  border: none;
  cursor: pointer;
  box-shadow: inset 1px 1px 0 0 rgba(255,255,255,.7),
    inset -1px -1px 0 0 rgba(255,255,255,.4), 1px 1px 1px 0 rgba(50,50,50,.1);

  &:active {
    @include opacity(.8);
  }

  .fa {
    margin-right: 6px;
  }
}

a.btn {
  &:hover,
  &:focus,
  &:active,
  &:visited {
    color: $g-text-btn;
    text-decoration: none;
  }
}

.btn-sm {
  padding: 4px 9px;
}

.btn-lg {
  padding: 9px 18px;
  font-size: 14px;
}

.btn-block {
  display: block;
}

.btn-default {
  background-color: #e8e8e8;
  background-image: linear-gradient(#e8e8e8,#e2e2e2);
  border: none;
  box-shadow: inset 1px 1px 0 0 rgba(255,255,255,.7),inset -1px -1px 0 0 rgba(255,255,255,.4),1px 1px 1px 0 rgba(50,50,50,.1);
}

.btn-primary {
  color: #fff;
  background-color: $g-brand-primary;
  background-image: none;
  border: none;
  //box-shadow: none;
}

.btn-success {
  color: #fff;
  background-color: $g-brand-success;
  background-image: none;
  //border:1px solid $g-brand-success;
}

.btn-info {
  color: #fff;
  background-color: $g-brand-info;
  background-image: linear-gradient($g-brand-info, $g-brand-info - #060606);
}
.btn-warning {
  color: #fff;
  background-color: $g-brand-warning;
  background-image: linear-gradient($g-brand-warning, $g-brand-warning - #060606);
}

.btn-danger {
  color: #fff;
  background-color: $g-brand-danger;
  background-image: linear-gradient($g-brand-danger, $g-brand-danger - #060606);
}


.btn-group {
  display: inline-block;
  margin: 0 5px;
  .btn {
    padding: 4px 12px;
    font-size: 12px;
    text-shadow: 1px 1px 0 rgba(255,255,255,.4);
    &:hover,
    &:focus,
    &:active,
    &:visited {
      color: #333;
    }
    &:last-child {
      border-right: none;
    }
    &.active {
      background-color: #c5c5c5;
      background-image: linear-gradient(#cbcbcb,#c5c5c5);
      box-shadow: inset 1px -1px 1px 0 rgba(255,255,255,.4),inset 1px 1px 1px 0 rgba(0,0,0,.2);
    }
  }
}
