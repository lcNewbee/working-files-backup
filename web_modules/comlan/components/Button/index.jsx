import './index.scss';
import React, {PropTypes} from 'react';
import Icon from '../Icon';

const propTypes = {
  iconName: PropTypes.string,
  role: PropTypes.string,
  className: PropTypes.string,
  size: PropTypes.oneOf(['sm', 'lg']),
  inverse: PropTypes.bool,
  Component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
};

const defaultProps = {
  Component: 'button',
};

class Button extends React.Component {

  render() {
    let { Component, role, size, className} = this.props;
    
    let classNames = `btn`;
    
    if (size) {
      classNames = `${classNames} btn-${size}`;
    }
    
    if (className) {
      classNames = className + ' ' + classNames;
    }
    
    return (
      <Component 
        {...this.props}
        className={classNames}
        type="button"
      >
        <Icon name={role} />
        {this.props.text}
      </Component>
    );
  }
}

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;