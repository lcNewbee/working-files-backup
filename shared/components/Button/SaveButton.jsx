import React, { PropTypes } from 'react';
import Button from './Button';

let timeOut = null;

const propTypes = {
  text: PropTypes.string,
  savingText: PropTypes.string,
  savedText: PropTypes.string,
  loading: PropTypes.bool,
};

const defaultProps = {
  text: _('Save'),
  savingText: _('Saving'),
  savedText: _('Saved'),
  theme: 'primary',
  icon: 'save',
};
class SaveButton extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      status: 'default',
    };
  }

  componentWillReceiveProps(nextProps) {
    // console.log('loading', nextProps.loading);
    if (this.props.loading && !nextProps.loading) {
      this.setState({
        status: 'ok',
      });

      timeOut = setTimeout(() => {
        this.setState({
          status: 'default',
        });
      }, 1000);
    } else if (nextProps.loading) {
      this.setState({
        status: 'saving',
      });
    }
  }

  componentWillUnmount() {
    clearTimeout(timeOut);
  }

  render() {
    const { text, savingText, savedText } = this.props;
    let showText = text;

    if (this.state.status === 'saving') {
      showText = savingText;
    } else if (this.state.status === 'ok') {
      showText = savedText;
    }

    return (
      <Button
        {...this.props}
        text={showText}
      />
    );
  }
}

SaveButton.propTypes = propTypes;
SaveButton.defaultProps = defaultProps;

export default SaveButton;
