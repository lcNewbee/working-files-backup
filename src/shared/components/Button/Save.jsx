import React, {PropTypes} from 'react';
import Button from './index';

let timeOut = null;

const propTypes = {
  text: PropTypes.string
};

const defaultProps = {
  text: _('Save'),
  role: 'primary',
  icon: 'save'
};

class SaveButton extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      status: 'default'
    };
  };

  componentWillReceiveProps(nextProps) {
    if(this.props.loading && !nextProps.loading) {
      this.setState({
        status: 'ok'
      });

      timeOut = setTimeout(function(){
        this.setState({
          status: 'default'
        });
      }.bind(this), 1000)
    } else if (nextProps.loading) {
      this.setState({
        status: 'saving'
      });
    }
  };

  componentWillUnmount() {
    clearTimeout(timeOut);
  };

  render() {
    let { text, loading, status} = this.props;

    if(this.state.status === 'saving') {
      text = _('Saving');
    } else if (this.state.status === 'ok') {
      text = _('Saved');
    }

    return (
      <Button
        {...this.props}
        text={text}
      />
    );
  }
}

SaveButton.propTypes = propTypes;
SaveButton.defaultProps = defaultProps;

export default SaveButton;
