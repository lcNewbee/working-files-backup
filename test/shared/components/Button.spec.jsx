// Link.react-test.js
import React from 'react';
import Button from 'shared/components/Button/Button';
import SaveButton from 'shared/components/Button/SaveButton';
import renderer from 'react-test-renderer';

describe('Shared component', () => {
  it('Button', () => {
    const component = renderer.create(
      <Button text="cog" />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    tree.props.onClick();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // // manually trigger the callback
    // tree.props.onMouseEnter();
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();

    // // manually trigger the callback
    // tree.props.onMouseLeave();
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();
  });

  it('SaveButton', () => {
    let component = renderer.create(
      <SaveButton text="cog" />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component = renderer.create(
      <SaveButton text="cog" loading />,
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // // manually trigger the callback
    // tree.props.onMouseEnter();
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();

    // // manually trigger the callback
    // tree.props.onMouseLeave();
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();
  });
});

