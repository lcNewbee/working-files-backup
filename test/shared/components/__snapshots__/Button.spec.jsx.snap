exports[`Shared component Button 1`] = `
<button
  className="a-btn a-btn--default"
  onClick={[Function]}
  role="button"
  type="button">
  cog
</button>
`;

exports[`Shared component Button 2`] = `
<button
  className="a-btn a-btn--default"
  onClick={[Function]}
  role="button"
  type="button">
  cog
</button>
`;

exports[`Shared component SaveButton 1`] = `
<button
  className="a-btn a-btn--primary"
  icon="save"
  onClick={[Function]}
  role="button"
  type="button">
  <span
    className="icon fa fa-save"
    name="save" />
  cog
</button>
`;

exports[`Shared component SaveButton 2`] = `
<button
  className="a-btn a-btn--primary a-btn--loading"
  icon="save"
  onClick={[Function]}
  role="button"
  type="button">
  <span
    className="icon fa fa-spinner fa-spin"
    name="spinner" />
  cog
</button>
`;
